### PR TITLE
Improve ClipRect documentation

### LINKED ISSUE #8557: Document ClipRect
It would be good to have docs that explain how to use a ClipRect.
Some questions that would be good to have answered:
* What widgets are used with a ClipRect(Transform maybe?) 
* What is a CustomClipper. 
* Are there prebuilt clippers?

Also an example sort of like what you have in the analogs:
Some like: How to a clip an image to only show the top half.

### PR DESCRIPTION
**Pull request by abarth**
Fixes #8557

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie @HansMuller

File: packages/flutter/lib/src/widgets/basic.dart [hunk 1]
```diff
@@ -252,6 +252,38 @@ class CustomPaint extends SingleChildRenderObjectWidget {
...
+///
+/// [ClipRect] is commonly used with these widgets, which can paint outside
+/// their bounds in some situations:
+///
+///  * [CustomPaint]
+///  * [CustomSingleChildLayout]
+///  * [CustomMultiChildLayout]
+///  * [Align] (e.g., if [Align.widthFactor] or [Align.heightFactor] is less
+///    than 1.0).
+///  * [OverflowBox]
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/basic.dart]: Maybe mention SizedOverflowBox despite the obvious name.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 2]
```diff
@@ -252,6 +252,38 @@ class CustomPaint extends SingleChildRenderObjectWidget {
...
 /// bounds, but the size and location of the clip rect can be customized using a
 /// custom [clipper].
+///
+/// [ClipRect] is commonly used with these widgets, which can paint outside
+/// their bounds in some situations:
+///
+///  * [CustomPaint]
+///  * [CustomSingleChildLayout]
+///  * [CustomMultiChildLayout]
+///  * [Align] (e.g., if [Align.widthFactor] or [Align.heightFactor] is less
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/basic.dart]: Maybe mention Center even though it's an Align subclass.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 3]
```diff
@@ -252,6 +252,38 @@ class CustomPaint extends SingleChildRenderObjectWidget {
 /// By default, [ClipRect] prevents its child from painting outside its
 /// bounds, but the size and location of the clip rect can be customized using a
 /// custom [clipper].
+///
+/// [ClipRect] is commonly used with these widgets, which can paint outside
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/basic.dart]: This seems a little subtle; we're not implying that all of the other widgets do clip. I can't think of a better way of phrasing the point though.

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 4]
```diff
@@ -881,6 +881,13 @@ class RenderBackdropFilter extends RenderProxyBox {
 /// supply a reclip argument to the constructor of the [CustomClipper]. The
 /// custom object will listen to this animation and update the clip whenever the
 /// animation ticks, avoiding both the build and layout phases of the pipeline.
+///
+/// See also:
+///
+///  * [ClipRect], which can be customized with a [CustomClipper].
+///  * [ClipRRect], which can be customized with a [CustomClipper].
+///  * [ClipOval], which can be customized with a [CustomClipper].
+///  * [ClipPath], which can be customized with a [CustomClipper].
```
Comment (HansMuller) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: Maybe mention Stack and any other widgets that do clip here?

**Comment by HansMuller** [comment]
LGTM

File: packages/flutter/lib/src/rendering/proxy_box.dart (hunk 4 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: I was mistaken about Stack.  It doesn't clip all the time.

File: packages/flutter/lib/src/widgets/basic.dart (hunk 1 above)
Comment (abarth) [inline packages/flutter/lib/src/widgets/basic.dart]: Done.

File: packages/flutter/lib/src/widgets/basic.dart (hunk 2 above)
Comment (abarth) [inline packages/flutter/lib/src/widgets/basic.dart]: Done.

**Pull request merged**